lib/ecto/parameterized_type.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Ecto.ParameterizedType do
+  @moduledoc """
+  Parameterized types are similar to Ecto Custom Types, however they allow a set of options to be
+  specified in the schema which are initialized on compilation and passed to the callback functions
+  as a final argument. Note that options should be specified as a keyword, but it is idiomatic to
+  convert them to maps inside init for easier pattern matching in other callbacks.
+
+  ## Examples
+
+    To create a parameterized type, create a module as shown below:
+
+    defmodule MyApp.MyType do
+      use Ecto.ParameterizedType
+
+      def type(_opts), do: :string
+
+      def init(opts) do
+        ...
+        Enum.into(opts, %{})
+      end
+
+      def cast(data, current, opts) do
```

Comment:
Instead of calling the third element `opts`, let's call it `config` or something else. `opts` kind of implies keyword. :)

Reply:
@josevalim Good point. It initially was always a keyword. Now that the standard will be to convert it to a map in init, opts doesn't work. How about `params` since it is a ParameterizedType?

Reply:
To me the params are the arguments given on `init` and not the inner state itself. That said, I am also fine with `params`. Anything besides `opts` is game for me. :)

Reply:
Sounds good. I am thinking that `opts` will remain the name of the keyword that is passed into init (including what we call it in schema as we are parsing the field macro), and params is the map that is returned from init, and everywhere that is used in subsequent callbacks. As an example, ParameterizedType would have these:

```elixir
  @type opts :: keyword()

  @type params :: map()

  @callback init(opts :: opts()) :: params()

  @callback type(params :: params()) :: Ecto.Type.t()
```

Reply:
Params should be any `term`. There is no reason for us to impose it being a map. :)